### PR TITLE
Add pseudo-index for key headers by height but dirtier

### PR DESCRIPTION
**Pull request by cytadela8**
Same as #3577, but using @gorbak25 direct rocksdb read.

### CODE REVIEW

**Comment by gorbak25** [comment]
image

File: apps/aecore/src/aec_db.erl [hunk 1]
```diff
@@ -521,13 +521,41 @@ find_key_headers_and_hash_at_height(Height) when is_integer(Height), Height >= 0
...
-            R = mnesia:dirty_select(aec_chain_state, [{#aec_chain_state{key = {key_header, Height, '_'},
-                                                                        value = '_'}
-                                                        , []
-                                                        , ['$_']
-                                                          }]),
-            [{Hash, aec_headers:from_db_header(Header)}
-             || #aec_chain_state{key = {key_header, _, Hash}, value = Header} <- R]
+            case persistent_term:get(?BYPASS, nobypass) of
+                {rocksdb, #{aec_chain_state := H}} ->
+                    WI = fun(Height, F) ->
```
Comment (gorbak25) [inline apps/aecore/src/aec_db.erl:526]: With iterator

File: apps/aecore/src/aec_db.erl [hunk 2]
```diff
@@ -521,13 +521,41 @@ find_key_headers_and_hash_at_height(Height) when is_integer(Height), Height >= 0
...
+                            {ok, I} = rocksdb:iterator(H,
+                                                       [{iterate_upper_bound,
+                                                         sext:prefix({key_header, Height+1, '_'})}]),
+                            try F(I) after rocksdb:iterator_close(I) end
+                        end,
+                    G = fun G(I, {ok, K ,V}, Height, Acc) ->
```
Comment (gorbak25) [inline apps/aecore/src/aec_db.erl:532]: Move to a module local function

**Comment by cytadela8** [comment]
@gorbak25 What's the speed difference?

**Comment by cytadela8** [comment]
~200ns vs ~300ns?

**Comment by gorbak25** [comment]
~200 us vs ~300 us

**Comment by cytadela8** [comment]
I have a dialyzer warning, I don't know how to fix:
```
apps/aecore/src/aec_db.erl
 579: The pattern 'stop' can never match the type {'cont',nonempty_maybe_improper_list()}
```

I think it complains about unused path? But I placed it there in case somebody wants to use it...

**Comment by uwiger** [comment]
> I think it complains about unused path? But I placed it there in case somebody wants to use it...

Yes, but since the function is local, and no call point produces that return value, dialyzer is unhappy.
Either export the function or comment out the case clause.

**Comment by cytadela8** [comment]
Rebased

File: apps/aecore/src/aec_chain_state.erl [hunk 3]
```diff
@@ -1347,3 +1349,78 @@ chain_ends_maybe_apply([H1|T1] = OldTops, [H2|T2], NewTops) -> %% Check if H1 is
...
+                done -> %% No migration is in progress
+                    case aec_db:find_key_headers_and_hash_at_height(1) of
+                        [] -> %% Ok this is an old DB - migration required
+                            aec_db:start_chain_migration(key_headers),
+                            start_key_headers_height_store_migration();
+                        _ -> %% No migration required
+                            ok
+                    end;
+                in_progress -> %% The migration was interrupted and not finished yet
+                    lager:info("[Orphan key blocks scan] Resuming interrupted scan"),
```
Comment (gorbak25) [inline apps/aecore/src/aec_chain_state.erl:1371]:
```suggestion
                    lager:info("[Key headers migrations scan] Resuming interrupted scan"),
```

File: apps/aecore/src/aec_chain_state.erl [hunk 4]
```diff
@@ -1347,3 +1349,78 @@ chain_ends_maybe_apply([H1|T1] = OldTops, [H2|T2], NewTops) -> %% Check if H1 is
...
+                                  10000,
+                                  read)
+                end),
+                {TotalTime, TotalCount} = key_headers_height_store_migration_step(Res),
+                lager:info("[Key headers migrations scan] DONE: In total migrated ~p headers in ~p microseconds", [TotalCount, TotalTime])
+            end),
+            aec_db:finish_chain_migration(key_headers)
+        catch
+            E:R:Stack ->
+                (catch io:format(user, "[Orphan key blocks scan] Terminating node: ~p ~p ~p", [E, R, Stack])),
```
Comment (gorbak25) [inline apps/aecore/src/aec_chain_state.erl:1397]:
```suggestion
                (catch io:format(user, "[Key headers migrations scan] Terminating node: ~p ~p ~p", [E, R, Stack])),
```

File: apps/aecore/src/aec_chain_state.erl [hunk 5]
```diff
@@ -1347,3 +1349,78 @@ chain_ends_maybe_apply([H1|T1] = OldTops, [H2|T2], NewTops) -> %% Check if H1 is
...
+                (catch lager:error("[Orphan key blocks scan] Terminating node: ~p ~p ~p", [E, R, Stack])),
```
Comment (gorbak25) [inline apps/aecore/src/aec_chain_state.erl:1398]:
```suggestion
                (catch lager:error("[Key headers migrations scan] Terminating node: ~p ~p ~p", [E, R, Stack])),
```

File: apps/aecore/src/aec_chain_state.erl [hunk 6]
```diff
@@ -1347,3 +1349,78 @@ chain_ends_maybe_apply([H1|T1] = OldTops, [H2|T2], NewTops) -> %% Check if H1 is
...
+                init:stop("[Orphan key blocks scan] Encountered a fatal error during the migration. Terminating the node")
```
Comment (gorbak25) [inline apps/aecore/src/aec_chain_state.erl:1399]:
```suggestion
                init:stop("[Key headers migrations scan] Encountered a fatal error during the migration. Terminating the node")
```

File: apps/aecore/src/aec_db.erl [hunk 7]
```diff
@@ -508,11 +511,74 @@ find_headers_and_hash_at_height(Height) when is_integer(Height), Height >= 0 ->
...
+        in_progress ->
+            R = mnesia:dirty_select(aec_headers, [{ #aec_headers{key = '_',
+                                                                 value = '$1',
+                                                                 height = Height}
+                                                  , [{'=:=', {element, 1, '$1'}, key_header}]
+                                                  , ['$_']}]),
+            [{Hash, aec_headers:from_db_header(Header)}
+             || #aec_headers{key = Hash, value = Header} <- R];
+        done ->
+            case persistent_term:get(?BYPASS, nobypass) of
```
Comment (gorbak25) [inline apps/aecore/src/aec_db.erl:524]: UNFEFINED BEHAVIOUR!!!
iterate_upper_bound is a optimization hint - not an obligation. 
Iterating past iterate_upper_bound is UB and anything might happen

**Review by gorbak25 — CHANGES_REQUESTED** [review]
UNFEFINED BEHAVIOUR!!!
iterate_upper_bound is a optimization hint - not an obligation.
Iterating past iterate_upper_bound is UB and anything might happen

File: apps/aecore/src/aec_db.erl (hunk 7 above)
Comment (cytadela8) [reply apps/aecore/src/aec_db.erl:524]: really? ok. Will fill

**Review by gorbak25 — COMMENTED** [review]
Before merging I will run some tests on a production DB to ensure everything works.

**Comment by gorbak25** [comment]
Ok, looks like it works fine - ~166 us per invocation :)

**Review by gorbak25 — APPROVED** [review]

**Review by velzevur — APPROVED** [review]

**Pull request merged**